--- src/network.c.orig	2012-03-10 02:11:29.000000000 +0100
+++ src/network.c	2012-03-10 02:13:01.000000000 +0100
@@ -705,7 +705,7 @@
 
 	exp_size = 3 * sizeof (uint16_t)
 		+ pkg_numval * (sizeof (uint8_t) + sizeof (value_t));
-	if ((buffer_len < 0) || (buffer_len < exp_size))
+	if (buffer_len < exp_size)
 	{
 		WARNING ("network plugin: parse_part_values: "
 				"Packet too short: "
@@ -790,7 +790,7 @@
 
 	uint16_t pkg_length;
 
-	if ((buffer_len < 0) || ((size_t) buffer_len < exp_size))
+	if ((size_t) buffer_len < exp_size)
 	{
 		WARNING ("network plugin: parse_part_number: "
 				"Packet too short: "
@@ -829,7 +829,7 @@
 
 	uint16_t pkg_length;
 
-	if ((buffer_len < 0) || (buffer_len < header_size))
+	if (buffer_len < header_size)
 	{
 		WARNING ("network plugin: parse_part_string: "
 				"Packet too short: "
